shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs

```diff
@@ -275,66 +278,264 @@ instance
+      (as a special case for empty asset names,
+       the index points to the end of the string)
+   D) a blob of policyIDs
+   E) a blob of asset names
+
+ The encoding strategy is
+   - Collect all (unique) policy Ids.
+   - Collect all (unique) asset names.
+   - Determine the sizes of the regions and allocate them.
+      - 12*numassets + len(concat(policyIds)) + len(concat(asset names))
+   - Write the policyIds to region D
+   - Write the asset names to region E
+   - For each asset entry
+     - Locate the corresponding asset name and policyId
+     - Write quantity, policyId location, and asset name location to
+       corresponding regions
+     - For the special case of 0 length asset names, the location is the
+       end of region E
+
+ The decoding strategy is
+   - Use length information to determine the beginnings of regions A,B,C
+     (We do not need to do this for regions D and E because the policyId
+      and asset name locations are relative to the beginning of entire rep.)
+   - For each integer in 0..(numassets -1)
+     - Read corresponding quantity, pid offset, and asset name offset from regions
+       A, B, and C, respectively.
+     - Read (pid length) bytes from pid offset. assume it points into region D
+     - Determine asset name lengths using the difference between the offset
+       and the next greater offset (if it exists). If there is no next greater
+       offset, use the difference from the end of the rep.
+       (Note: for the special case of 0 length asset names, this calculation
+       results in 0 because we are subtracting the end of the rep from itself.)
+     - Read (asset name length) bytes from asset name offset. assume it points
+       into region E.
+ -}
   deriving (Eq, Show, Typeable)
 
 deriving via
-  OnlyCheckWhnfNamed "CVPart" (CVPart crypto)
+  OnlyCheckWhnfNamed "CompactValue" (CompactValue crypto)
   instance
-    NoThunks (CVPart crypto)
-
-toCV :: Value crypto -> Maybe (CV crypto)
-toCV v = do
-  let (c, triples) = gettriples v
-      policyIDs = Set.fromList $ (\(x, _, _) -> x) <$> triples
-      n = length triples - 1
-  cvParts <- traverse (toCVPart policyIDs) triples
-  let arr = array (0, n) (zip [0 .. n] cvParts)
-  c' <- integerToWord64 c
-  pure $ CV c' arr
+    NoThunks (CompactValue crypto)
+
+to ::
+  forall crypto.
+  (CC.Crypto crypto) =>
+  Value crypto ->
+ -- The Nothing case of the return value corresponds to a quantity that is outside
+ -- the bounds of a Word64. x < 0 or x > (2^64 - 1)
+  Maybe (CompactValue crypto)
+to v = do
+  c <- integerToWord64 ada
+  -- Here we convert the (pid, assetName, quantity) triples into
+  -- (Int, (Word16,Word16,Word64))
+  -- These represent the index, pid offset, asset name offset, and quantity.
+  -- If any of the quantities out of bounds, this will produce Nothing.
+  -- The triples are ordered so that a component with an asset name appearing
+  -- earlier in the asset name section is first.
```

Comment:
Is this ordering stable? What happens with multiple null asset names, which will share the offset? Can this result in different representations, and do we care?